### ISSUE TITLE
Replacement of mplayer, upgrade to gcc6 and updating of many deps.

### ISSUE THREAD

**Issue by DeadSix27** [rdp/ffmpeg-windows-build-helpers#211]
^Began that.

Yet a few issues to solve tho:

- [x]  ~~mpv + vs R36 seems to not work properly (r35 works) ~~(waiting for response on a issue post)~~ Still unsure, rebuilding everything again, dude said it might be my python setup, maybe wrong gendef borks it? in that case.. lets..~~
fixed.
- [x] ~~..build mingw64 tools as well in the mingw script (need that for gendef) (thankfully just a argument, was easy to add)~~
- [x] ~~autopatch weak ref~~
- [ ] fix/add libarchive support in mpv?

libarchive with xml2
```
libarchive fails to build:
libxml2.a(encoding.o):encoding.c:(.text+0x2ccc): undefined reference to `libiconv'
```
without:
```
mpv fails to build:
libxml2.a(xzlib.o):xzlib.c:(.text+0xdac): undefined reference to `lzma_auto_decoder'
```
- [ ] finish my python/vapoursynth library script (which generates the libraries for mingw)
- [ ] decide whether i care enough about OpenAL in mpv to write downloader for: https://github.com/meganz/mingw-std-threads, requires a little patching in openal )
- [ ] Update libbluray?
- [x] Not to mention this issue `undefined reference to `__imp_EnumFontFamiliesExW'` when linking libbluray in ffmpeg
Maybe -lgdi32 could fix it (yep it did)
- [ ] make it easy to write some kind of plugins, as I'd like to use this toolchain with a few other things as well (and not have multiple.. or various annoying scripts)
- [x] convince @rdp that build_libtheora can go and won't be missed ~~(it prevents me from updating a few other deps..) or find a patch for it.
Someone with more c++ knowledge than I could possible take a look at it.~~
**Actually still works with a small patch i made, (for now)**
- [ ] get rid of mplayer (in favour of mpv) (which means libdvdnav/read can be updated (woo)
- [X] and the biggest of all: go to GCC 6.3.0 which works fine with _GLIBCXX_USE_WEAK_REF 0
- [ ] strip VLC and deps, doubt it works still (Qt especially is hard to compile I highly doubt that works)
- [ ] also find out what causes:
```
po_FileClose.c:44:20: fatal error: pocore.h: No such file or directory
po_atof.c:42:28: fatal error: PceNativeCall.h: No such file or directory
flop.c:55:24: fatal error: palm_flite.h: No such file or directory
```
in flite..

- [ ] write a small function to strip out the cross-compile stuff and prefix in ffmpeg's version output (like zeranoe does)

Basically, all works fine on GCC 6.3.0 with one patch and the ditch of libtheora.

**Comment by rdp** [comment]
what is libtheora used for, is that ogg?  Yeah you guessed it as to why I
haven't updated libbluray LOL [java ME? whoa forget it! LOL].  I've tried
submitting some complaints to the libtheora people in the past wonder what
happened...or if it's dead LOL.  thanks for your feedback to drive things
forward...

**Comment by DeadSix27** [comment]
@rdp ikr that java crap is annoying, i wonder if we can just supply it with the jar? It compiles just fine, even theora, i switched to git on most of the xiph stuff, the git seems to have all the fixes for which we had to use patches... so far it compiles fine with mingw from git, gcc 6.3.0 and a big amount of updated deps.

**Comment by DeadSix27** [comment]
@rdp oops forgot theora, ye that actually seems to work now with a single patch and using git master, (which didnt get updated since 2016.. so i guess its stable enough)
all in all i have like 3 new patches  and about 10-12 updated deps. minus t he upcoming mpv stuff, which i will do later (after getting a working gcc6.3.0/mingwgit setup going)

**Comment by rdp** [comment]
sounds fine here...except that patched gcc I've been reluctant to pull the
trigger there since I call out to a separate script for it hmmm...

**Comment by DeadSix27** [comment]
@rdp ye, i modified the seperate script..:

```
	if [ "$gcc_ver" == "6.3.0" ]; then # weak ref patch, see https://github.com/Alexpux/MINGW-packages/issues/1580
		cd gcc-"$gcc_ver"
			echo "Patching GCC 6.3.0"
			curl -4 --retry 5 "https://dsix.tech/gcc_6_3_0_weak_refs_x86_64.patch" -O --fail || exit 1 # don't forget to replace my serv with github
			echo "applying patch"
			patch -p1 < "gcc_6_3_0_weak_refs_x86_64.patch"
			echo "Done"
		cd ..
	fi
```

**Comment by DeadSix27** [comment]
@rdp any idea how to solve this?:
https://forum.qt.io/topic/77176/pkg-config-problem-when-cross-compiling-qt-5-8-linux-windows-mingw

thats really the only thing stopping it from being a fully working qt compilation.. :/

**Comment by rdp** [comment]
You can manually adjust .pc files FWIW...

**Comment by DeadSix27** [comment]
@rdp ye...........................

Manuallllyyy.....
(2hours later) = 
```
import os,re,sys
PATH = sys.argv[1]
flist = [os.path.join(dp, f) for dp, dn, filenames in os.walk(PATH) for f in filenames if os.path.splitext(f)[1] == '.prl']

for f in flist:
	path = f
	with open(path, "r+") as file:
		pathContentOriginal = file.read()
		
		m = re.search(re.compile(r"([ ])(\/)(lib[A-Za-z0-9^\S]+)(\.a)", re.MULTILINE),pathContentOriginal)
		if m != None:
			print(m.groups())
			pathContentFixed = re.sub(r"([ ])(\/)(lib[A-Za-z0-9^\S]+)(\.a)", r"\1\3" , pathContentOriginal)
			print("fixing "+ path)
			print(pathContentFixed)
			file.seek(0)
			file.truncate()
			file.write(pathContentFixed)
```

**Comment by DeadSix27** [comment]
@rdp https://github.com/DeadSix27/modular_cross_compile_script

**Comment by rdp** [comment]
Shoot I was expecting it to just be like an import-able diff, oh well.
Good luck with your project!

**Comment by DeadSix27** [comment]
@rdp Well, I did update a lot of deps in your script too, that project is just a side thing.

**Comment by DeadSix27** [comment]
@rdp 
I pretty much updated all the deps now, should be easy to "backport" patches and other things from my script to yours.
Ill do that next.
However, i run into one issue, with all those updated deps..
ffmpeg STATIC works fine, but SHARED doesn't compile.. did you ever encounter this problem before:

```
cmdutils.o: In function `print_device_sources':
cmdutils.c:2141: undefined reference to `avdevice_list_input_sources'
cmdutils.c:2141:(.text+0x1dd): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `avdevice_list_input_sources'
cmdutils.c:2152: undefined reference to `avdevice_free_list_devices'
cmdutils.c:2152:(.text+0x246): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `avdevice_free_list_devices'
cmdutils.o: In function `print_device_sinks':

```

I stripped the shared build and all that's left is:
```python
' --arch={bit_name2}'
' --target-os=mingw32'
' --cross-prefix={cross_prefix_bare}'
' --pkg-config=pkg-config'
' --disable-w32threads'
' --prefix={product_prefix}/ffmpeg_shared_git.installed'
' --enable-shared'
' --disable-static'
' --disable-libgme' #actually think this could work now in shared too.
```
Yet it still complains:

**Comment by rdp** [comment]
what if you adjust the ffmpeg configure and remove the image-base thing?
https://trac.ffmpeg.org/ticket/5895

**Comment by DeadSix27** [comment]
Doesn't resolve it unfortunately.

However after much narrowing I found out its SDL1 causing the issue..
that dumb CFLAG
`'custom_cflag' : '-DDECLSPEC=',# avoid SDL trac tickets 939 and 282, and not worried about optimizing yet...`

I had forgotten setting that.. (ofc that issue is still a thing too... You can't imagine how many work-arounds and bugs still exist, even with so many deps updated..) \*rolls eyes\*

**Comment by DeadSix27** [comment]
@rdp Also,
What causes ffmpeg to make it "private" and not public
The wrongly generated pc files cause other programs to fail linking any idea?
Incorrect (how it generates for me):
```pkgconfig
Requires: 
Requires.private: libavcodec >= 57.86.103, libswresample >= 2.4.100, libavutil >= 55.51.100
Conflicts:
Libs: -L${libdir}  -lavformat 
Libs.private: -lws2_32 -lstdc++ -lole32 -loleaut32 -lavicap32 -lgdi32 -lpsapi -lole32 -lstrmiids 
```
And not like this? :
Correct (how it should):
```pkgconfig
Requires: libavcodec >= 57.82.102, libswresample >= 2.4.100, libavutil >= 55.48.10
Requires.private: 
Conflicts:
Libs: -L${libdir}  -lavformat -lws2_32 -ldl -ldl -ldl -lstdc++ -lole32 -loleaut32 
Libs.private: 
```

**Comment by rdp** [comment]
this is a qwirk with pkg-config where if you say "pkg-config --static" or something it gives you the right stuff.  But almost no build system has a way to do that that I'm aware of (except maybe VLC) so I end up just modifying .pc files...

**Comment by DeadSix27** [comment]
@rdp Good to know, ye, but in this case I was just blind and didn't notice I built ffmpeg shared rather than static which it has to be for all the programs we/I are/am linking to.

**Comment by rdp** [comment]
I think private is for "static" libraries and normal Libs is for "shared"
equivalents...

**Comment by DeadSix27** [comment]
@rdp indeed.

**Comment by DeadSix27** [comment]
@rdp https://github.com/DeadSix27/modular_cross_compile_script/tree/dev
All libraries are as up to date as possible and have correct patches feel free to take them.
Ill slowly work on updating the libs in your script too, but together its obviously faster.
But those are only tested on ubuntu 16.10-64 and GCC6.3.0 with the weak refs patched.

I was going to test on OpenSUSE later this week (maybe weekend) and see if it compiles just fine.
Either way, you should at least be able to take most of the work out of it.

For example, even libbluray. (without java tho :P)

**Comment by hydra3333** [comment]
> I'll slowly work on updating the libs in your script too, but together its obviously faster.

How's it going ?

I was going through updating stuff (including adding draft nvidia OpenCL updates) like mentioned here https://github.com/rdp/ffmpeg-windows-build-helpers/issues/219#issuecomment-292953744 and then thought of going to gcc 6.3.0 and whatnot and saw your good work here ... just after I'd made a local copy of mingw-w64-build-3.6.7.local and was messing around with it 

```
mingw_w64_release_ver='5.0.2' # or git :) #mingw_w64_ver='git'
gcc_release_ver='6.3.0' # 6.1.0 has problems with 64 bit ffmpeg with c++ deps
gcc_old_release_ver='4.9.4' # default if none specified, was 4.9.4 cannot be same as gcc_release_ver above, apparently, as a note...
mpfr_release_ver='3.1.5'
mpc_release_ver='1.0.3'
binutils_release_ver='2.28'
gmp_release_ver='6.1.2'
isl_release_ver='0.16.1'
pthreads_w32_release_ver='2-9-1'
```
I'd give up and wait. if you were still actively engaged with it :)

**Comment by DeadSix27** [comment]
@hydra3333 libs are all updated (On my script) I haven't really updated rdp's ones yet.. then again my code is open too so you could just port them over to rdp too:
https://github.com/DeadSix27/modular_cross_compile_script

**Comment by hydra3333** [comment]
Thanks, yes I've done a bit of it (thanks !) but would dally about the whole on the basis you mentioned you were going to do it :)  Python isn't my strong point (stronger than building anything with minw/gcc which tells you something.

My scripted links to things like these just broke, so I guess you're making changes and so I'll need to update those links  
curl --retry 5 "https://raw.githubusercontent.com/DeadSix27/modular_cross_compile_script/master/patches/mingw/0003-dxgi-Add-missing-dxgi-1.2-structs-and-interfaces.patch"

**Comment by DeadSix27** [comment]
@hydra3333 oh ye i restructured some things and the git thing is just so it always uses git.

**Comment by hydra3333** [comment]
Oh joy, you're hostingthe Opus patch
`https://raw.githubusercontent.com/DeadSix27/modular_cross_compile_script/master/patches/opus-1.1.2.patch`

which still works with git version
```
  do_git_checkout https://github.com/xiph/opus.git
  cd opus_git
```

**Comment by DeadSix27** [comment]
@hydra3333 
```python
	'libopus' : {
		'repo_type' : 'archive',
		'url' : 'https://github.com/xiph/opus/releases/download/v1.1.2/opus-1.1.2.tar.gz',
		'patches': (
			("https://raw.githubusercontent.com/DeadSix27/modular_cross_compile_script/master/patches/opus-1.1.2.patch", "p0"),
		),
		'configure_options': '--host={compile_target} --prefix={compile_prefix} --disable-shared --enable-static',
		'make_options': '',
	},
```
Just look at my source if you fail to build a library, but note, those are only tested working on GCC6.3.0 and posix threads.

**Comment by hydra3333** [comment]
:) Thanks.  I have been perusing.

**Comment by hydra3333** [comment]
Ah, https://github.com/DeadSix27/modular_cross_compile_script/issues/5#issuecomment-293022252
> libfdk_aac  is not included, its not allowed in a GPL licensed build. So I just didn't include it at all. its very easy to add yourself tho, you can do a pull request if you implemented it if you want. (so at least the code is there and people only have to add that to the dependency)
>  I do not use FDK-AAC (nonGPL)

no non-GPL means I roll my own.   oh well, all good.  when the script is stable i'll look at adding any missing stuff I use. Although I'm not as yet quite sure what to do about the sets of OpenCL actions.

@DeadSix27 out of interest (lazy me)
- do you do an equivalent to rdp's libav build thing, so that x264 gets (re)built with libav so that it can handle mpeg containers and whatnot as direct input ?
- could you also do an equivalent to this, which gets done after "build_apps" finishes in rdp's script ?  It results in x264 also being able to create output in a .mp4 container which can be handy -
```
  if [[ $build_lsw == 'y' ]]; then
     #build_lsw
	 rm -rf x264
	 rm -rf x264_normal_bitdepth
	 rm -rf x264_high_bitdepth_10
	 rm -rf x264_with_libav_normal_bitdepth
	 rm -rf x264_with_libav_high_bitdepth_10
     build_libx264
  fi
```

**Comment by DeadSix27** [comment]
@rdp I build x264 only with libAV, no 10bit, no libav-less one, x264-8bit-libav is all there is.
I do not need any other one, for 10bit I compress in x265.
Yet again, its very simple for persons to add that to my script, its like, 2 lines to edit.

**Comment by hydra3333** [comment]
> Oh joy, you're hosting the Opus patch
> https://raw.githubusercontent.com/DeadSix27/modular_cross_compile_script/master/patches/opus-1.1.2.patch
> which still works with git version
>   do_git_checkout https://github.com/xiph/opus.git
>   cd opus_git

I spoke a tad too soon.  That patch doesn't work with the latest git, you need one like this
[libopus.patch.txt](https://github.com/rdp/ffmpeg-windows-build-helpers/files/915646/libopus.patch.txt)

**Comment by DeadSix27** [comment]
As I instead of adding mpv to this build, added it to my own thing as well as someone of you now upgrades the libraries, I should probably close this
